### PR TITLE
[CIAC-10175] Reimplement Demisto Lock to use polling mechanism

### PR DESCRIPTION
**Pull request by samuelFain**
## Contributing to Cortex XSOAR Content
Make sure to register your contribution by filling the [contribution registration form](https://forms.gle/XDfxU4E61ZwEESSMA)

**The Pull Request will be reviewed only after the contribution registration form is filled.**

## Status
- [ ] In Progress
- [x] Ready
- [ ] In Hold - (Reason for hold)

## Related Issues
fixes: [CIAC-10175](https://jira-dc.paloaltonetworks.com/browse/CIAC-10175)

## Description
- Reimplement Demisto Lock to use polling mechanism.
- Update TPB to run on both XSOR 6.x.x and XSOAR 8.x.x 

### TPB run successfully
* "Check Timeout" task will continue on error on XSOAR 8.x.x
DemistoLockTest_Wed_May_22_2024

### CODE REVIEW

File: Packs/DemistoLocking/Integrations/DemistoLock/DemistoLock.js [hunk 1]
```diff
@@ -57,33 +56,63 @@ switch (command) {
             logDebug('Task guid: ' + guid + ', Incident:' + incidentID + ' | Current lock is: ' + JSON.stringify(lock) + ', version: ' + version);
 
             if (lock.guid === guid) {
-                break;
+                var md = '### Demisto Locking Mechanism\n';
```
Comment (JasBeilin) [inline Packs/DemistoLocking/Integrations/DemistoLock/DemistoLock.js:59]: Lets add debug log.

File: Packs/DemistoLocking/Integrations/DemistoLock/DemistoLock.js [hunk 2]
```diff
@@ -57,33 +56,63 @@ switch (command) {
...
+                    try {
+                        setLock(guid, lockInfo, version);
+                    } catch (err) {
+                        logDebug(err.message)
+                    }
+                }
+                var timeout_err_msg = 'Timeout waiting for lock\n';
+                timeout_err_msg += 'Lock name: ' + lockName + '\n';
+                timeout_err_msg += 'Lock info: ' + lock.info + '\n';
+                return {
```
Comment (JasBeilin) [inline Packs/DemistoLocking/Integrations/DemistoLock/DemistoLock.js:76]: Lets add debug logs

File: Packs/DemistoLocking/Integrations/DemistoLock/DemistoLock.js [hunk 3]
```diff
@@ -39,16 +39,15 @@ switch (command) {
...
 
-        var guid = guid();
+        var guid = args.guid || guid();
         var time = 0;
         var lock, version;
-        var attempt = 1;
-        do{
-            logDebug('Task guid: ' + guid + ', Incident:' + incidentID + ' | Trying to acquire lock lockName: ' + lockName + ', attempt number: ' + attempt);
+
+        if (isDemistoVersionGE('8.0.0')) {  // XSOAR 8 lock implementation with polling.
```
Comment (JasBeilin) [inline Packs/DemistoLocking/Integrations/DemistoLock/DemistoLock.js:49]: Lets log the version recieved.

File: Packs/DemistoLocking/Integrations/DemistoLock/DemistoLock.js [hunk 4]
```diff
@@ -57,33 +56,63 @@ switch (command) {
...
             }
-            if (!lock.guid) {
-                try {
-                    setLock(guid, lockInfo, version);
-                } catch(err) {
-                    logDebug('Task guid: ' + guid + ', Incident:' + incidentID + ' | Failed setting lock: ' + err.message);
+            else {
+                // attempt to acquire the lock
+                if (!lock.guid) {
+                    try {
```
Comment (JasBeilin) [inline Packs/DemistoLocking/Integrations/DemistoLock/DemistoLock.js:67]: Lets add debug log.

File: Packs/DemistoLocking/Integrations/DemistoLock/DemistoLock.js [hunk 5]
```diff
@@ -57,33 +56,63 @@ switch (command) {
...
+                }
+                var timeout_err_msg = 'Timeout waiting for lock\n';
+                timeout_err_msg += 'Lock name: ' + lockName + '\n';
+                timeout_err_msg += 'Lock info: ' + lock.info + '\n';
+                return {
+                    Type: entryTypes.note,
+                    Contents: 'Lock was not acquired, Polling.',
+                    PollingCommand: 'demisto-lock-get',
+                    NextRun: '30',
+                    PollingArgs: { name: lockName, info: args.info, timeout: args.timeout, guid: guid, timeout_err_msg: timeout_err_msg },
```
Comment (JasBeilin) [inline Packs/DemistoLocking/Integrations/DemistoLock/DemistoLock.js:81]: Sending the custom message won't break polling?

**Review by JasBeilin — CHANGES_REQUESTED** [review]
looks good, please add much more logging so our devs can handle issues in all versions.
Lets also add documentation regarding the way it works.

File: Packs/DemistoLocking/Integrations/DemistoLock/DemistoLock.js (hunk 5 above)
Comment (samuelFain) [reply Packs/DemistoLocking/Integrations/DemistoLock/DemistoLock.js:81]: No, it works as expected. The `timeout_err_msg` argument is a preparation for overriding the default timeout message.

**Comment by samuelFain** [comment]
@JasBeilin 
Added relevant debug logs as suggested.
Also created new docs for this integration with all relevant information [here](https://confluence-dc.paloaltonetworks.com/display/DemistoContent/Demisto+Lock).

**Review by JasBeilin — APPROVED** [review]
Nice

**Pull request merged**